compiler/rustc_lint_defs/src/lib.rs

```diff
@@ -46,13 +48,60 @@ pub enum Applicability {
     Unspecified,
 }
 
+rustc_index::newtype_index! {
+    /// FIXME: The lint expectation ID is currently a simple copy of the `AttrId`
+    /// that the expectation originated from. In the future it should be generated
+    /// by other means. This is for one to keep the IDs independent of each other
+    /// and also to ensure that it is actually stable between compilation sessions.
+    /// (The `AttrId` for instance, is not stable).
```

Comment:
The explanation is a bit more complicated, but here is the gist of it:

From my understanding, this is required for two reasons:
1. The incremental compilation implementation stores diagnostics between compilations on disc and replays them if the producing query has been marked as green (meaning that it won't be executed). This means that a diagnostic can be emitted with a `LintExpectationId` that originated from the last session. (I'm not totally sure on the rules, it might be that this is not actually a problem. The question would be if any of the linting query can be marked as green when any expectations have been changed for that crate/module)
2. The expectations are collected during the [`lint_levels`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.lint_levels) query and stored in the [`LintLevelMap`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/lint/struct.LintLevelMap.html) struct. The query is marked as `eval_always` but my understanding is that the result is then hashed and compared with the stored hash. The `LintExpectationId` is part of the query result and therefore also used to check if the query result has changed. Having it unstable can result in a different hash for the same query input.